tests/unit_tests/subaddress.cpp

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2014-2017, The Monero Project
+// 
+// All rights reserved.
+// 
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+// 
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+// 
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+// 
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+// 
+// Parts of this file are originally copyright (c) 2012-2013 The Cryptonote developers
+#include <boost/filesystem.hpp>
+#include "gtest/gtest.h"
+
+#include "include_base_utils.h"
+#include "wallet/wallet2.h"
+#include "crypto/crypto.h"
+#include "cryptonote_basic/account.h"
+#include "cryptonote_basic/cryptonote_basic_impl.h"
+#include "wallet/api/subaddress.h"
+
+class WalletSubaddress : public ::testing::Test 
+{
+  protected:
+    virtual void SetUp() 
+    {
+      try
+      {
+          w1.generate(wallet_name, password, recovery_key, true, false);
+      }
+      catch (const std::exception& e)
+      {
+          LOG_ERROR("failed to generate wallet: " << e.what());
+      }
```

Comment:
A test ought to fail on errors.

Reply:
Definitely, wasn't sure of best way to fail out. Should I just use `exit(1)`?

Reply:
I believe gtest will fail the test on uncaught exception. Do check though, I'm not 100% sure. Otherwise, do whatever other tests do.

Reply:
In `unit_tests/serialization.cpp` the test just uses the try-catch block with an empty exception handler. So, I think logging the error may be enough, and the exception will bubble up?

In `functional_tests/transaction_flow_test.cpp` the try-catch block returns false on exception, but that is because the test harness is a function that returns a bool.

Only way to really know is cause an intentional error in wallet generation. Will providing improper input to the `generate` function get it past the compiler, and still cause the exception we're looking for?

Reply:
throw runtime_error("");

Reply:
Thanks @moneromooo-monero just added the throw to the test. Included `e.what()` for extra detail, can remove if this is unnecessary.